commitrmnocb/observation.go

```diff
@@ -0,0 +1,171 @@
+package commitrmnocb
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"sort"
+
+	"github.com/smartcontractkit/libocr/offchainreporting2plus/ocr3types"
+	"github.com/smartcontractkit/libocr/offchainreporting2plus/types"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/hashutil"
+	"github.com/smartcontractkit/chainlink-common/pkg/merklemulti"
+	cciptypes "github.com/smartcontractkit/chainlink-common/pkg/types/ccipocr3"
+
+	"github.com/smartcontractkit/chainlink-ccip/plugintypes"
+)
+
+func (p *Plugin) ObservationQuorum(_ ocr3types.OutcomeContext, _ types.Query) (ocr3types.Quorum, error) {
+	// Across all chains we require at least 2F+1 observations.
+	return ocr3types.QuorumTwoFPlusOne, nil
+}
+
+func (p *Plugin) Observation(
+	ctx context.Context, outCtx ocr3types.OutcomeContext, _ types.Query,
+) (types.Observation, error) {
+	previousOutcome, nextState := p.decodeOutcome(outCtx.PreviousOutcome)
+
+	observation := Observation{}
+	switch nextState {
+	case SelectingRangesForReport:
+		offRampNextSeqNums := p.ObserveOffRampNextSeqNums(ctx)
+		observation = Observation{
+			OnRampMaxSeqNums:   offRampNextSeqNums,
+			OffRampNextSeqNums: offRampNextSeqNums,
+			FChain:             p.ObserveFChain(),
+		}
+
+	case BuildingReport:
+		observation = Observation{
+			MerkleRoots: p.ObserveMerkleRoots(ctx, previousOutcome.RangesSelectedForReport),
+			GasPrices:   []cciptypes.GasPriceChain{},
+			TokenPrices: []cciptypes.TokenPrice{},
+			FChain:      p.ObserveFChain(),
+		}
+
+	case WaitingForReportTransmission:
+		observation = Observation{
+			OffRampNextSeqNums: p.ObserveOffRampNextSeqNums(ctx),
+			FChain:             p.ObserveFChain(),
+		}
+
+	default:
+		p.lggr.Warnw("Unexpected state", "state", nextState)
+		return observation.Encode()
+	}
+
+	p.lggr.Infow("Observation", "observation", observation)
+	return observation.Encode()
+}
+
+// ObserveOffRampNextSeqNums observes the next sequence numbers for each source chain from the OffRamp
+func (p *Plugin) ObserveOffRampNextSeqNums(ctx context.Context) []plugintypes.SeqNumChain {
+	supportsDestChain, err := p.supportsDestChain(p.nodeID)
+	if err != nil {
+		p.lggr.Warnw("call to SupportsDestChain failed", "err", err)
+		return nil
+	}
+
+	if supportsDestChain {
+		sourceChains := p.knownSourceChainsSlice()
+		offRampNextSeqNums, err := p.ccipReader.NextSeqNum(ctx, sourceChains)
+		if err != nil {
+			p.lggr.Warnw("call to NextSeqNum failed", "err", err)
+			return nil
+		}
+
+		if len(offRampNextSeqNums) != len(sourceChains) {
+			p.lggr.Warnf("call to NextSeqNum returned unexpected number of seq nums, got %d, expected %d",
+				len(offRampNextSeqNums), len(sourceChains))
+			return nil
+		}
```

Comment:
We're tolerant to an unexpected number of next seq nums, though this error make sense as an "error" instead of a "warn".

>  Is it possible to avoid that dependency with a more tolerant check here?

Possibly, let's think about this in future PRs.